tests/phpunit/tests/dependencies/scripts.php

```diff
@@ -63,6 +66,699 @@ public function test_wp_enqueue_script() {
+		$output = get_echo( 'wp_print_scripts' );
+
+		$expected  = "<script type='text/javascript' src='http://example.org/ds-is2-1.js' id='ds-is2-1-js' defer></script>\n";
+		$expected .= "<script type='text/javascript' id='ds-is2-2-js-before'>\n";
+		$expected .= "console.log(\"before one\");\n";
+		$expected .= "</script>\n";
+		$expected .= "<script type='text/javascript' src='http://example.org/ds-is2-2.js' id='ds-is2-2-js' defer></script>\n";
+		$expected .= "<script type='text/javascript' src='http://example.org/ds-is2-3.js' id='ds-is2-3-js' defer></script>\n";
+		$expected .= "<script type='text/javascript' src='http://example.org/ms-is2-1.js' id='ms-is2-1-js' defer></script>\n";
+
+		$this->assertSame( $expected, $output );
+	}
+
+	/**
+	 * Test valid async loading strategy case.
+	 *
+	 * @ticket 12009
+	 */
+	public function test_loading_strategy_with_valid_async_registration() {
+		// No dependents, No dependencies then async.
+		wp_enqueue_script( 'main-script-a1', '/main-script-a1.js', array(), null, array( 'strategy' => 'async' ) );
+		$output   = get_echo( 'wp_print_scripts' );
+		$expected = "<script type='text/javascript' src='/main-script-a1.js' id='main-script-a1-js' async></script>\n";
+		$this->assertSame( $expected, $output );
+	}
+
+	/**
+	 * Test invalid async loading strategy cases.
+	 *
+	 * @ticket 12009
+	 */
+	public function test_loading_strategy_with_invalid_async_registration() {
+		// If any dependencies then it's not async. Since dependency is blocking(/defer) final strategy will be defer.
+		wp_enqueue_script( 'dependency-script-a2', '/dependency-script-a2.js', array(), null );
+		wp_enqueue_script( 'main-script-a2', '/main-script-a2.js', array( 'dependency-script-a2' ), null, array( 'strategy' => 'async' ) );
+		$output   = get_echo( 'wp_print_scripts' );
+		$expected = "<script type='text/javascript' src='/main-script-a2.js' id='main-script-a2-js' defer></script>";
+		$this->assertStringContainsString( $expected, $output, 'Expected defer.' );
+
+		// If any dependent then it's not async. Since dependent is not set to defer the final strategy will be blocking.
+		wp_enqueue_script( 'main-script-a3', '/main-script-a3.js', array(), null, array( 'strategy' => 'async' ) );
+		wp_enqueue_script( 'dependent-script-a3', '/dependent-script-a3.js', array( 'main-script-a3' ), null );
+		$output   = get_echo( 'wp_print_scripts' );
+		$expected = "<script type='text/javascript' src='/main-script-a3.js' id='main-script-a3-js'></script>";
+		$this->assertStringContainsString( $expected, $output, 'Expected blocking.' );
+	}
+
+	/**
+	 * Test valid defer loading strategy cases.
+	 *
+	 * @ticket 12009
+	 * @dataProvider data_loading_strategy_with_valid_defer_registration
+	 */
+	public function test_loading_strategy_with_valid_defer_registration( $expected, $output, $message ) {
+		$this->assertStringContainsString( $expected, $output, $message );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
```

Comment:
```suggestion
	 * @return array[]
```
The agreed "for now" docblock for data providers is below, until the community reach consensus for an alternative.

```php
/**
 * Data provider.
 *
 * @return array[]
 */
```